source/common/network/udp_listener_impl.cc

```diff
@@ -34,20 +34,49 @@ UdpListenerImpl::UdpListenerImpl(Event::Dispatcher& dispatcher, SocketSharedPtr
     : BaseListenerImpl(dispatcher, std::move(socket)), cb_(cb), time_source_(time_source),
       // Default prefer_gro to false for downstream server traffic.
       config_(config, false) {
+  parent_drained_callback_registry_ =
+      dynamic_cast<const UdpListenSocket&>(*socket_).parentDrainedCallbackRegistry();
```

Comment:
Why is a dynamic_cast needed? Can the types be changed so that this isn't needed?

Reply:
Couldn't upgrade the types because it cascaded back up through multiple layers to eventually interfaces which, in order to be able to use the specified type, caused a circular dependency.
So instead I put a virtual function on the base Socket type, which is yuck, but works cleanly and doesn't cause any kind of dependency cascade.